designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,556 @@
+- Feature Name: Latent Data Block
+- Start Date: 09-30-2024
+
+# Summary
+[summary]: #summary
+
+This design doc proposes adding a new block to the Stan language
+called `latent data` with a range of proposed uses.  The new block
+will appear between the transformed parameter and model blocks and be
+executed once at the beginning of each iteration using `double` values
+for all parameters and transformed parameters.  The `latent data`
+variables will then be in scope for the rest of the program, but
+cannot be modified.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation is to support three new key features for Stan:
+
+* Cut-based reasoning for random quantities 
+* Multiple imputation of missing data
+* Gibbs sampling discrete parameters
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Stan will support a new block, `latent data`, that is required to
+appear between `transformed parameters` and the `model` block.
```

Comment:
Thanks---I like that name.  It gets confusing because a lot of the missing data stuff is done in a frequentist setting where they treat it as latent data.

I don't think we'll need Jacobian adjustments for continuous parameters based on discrete parameters.  There is a major issue in sequencing these things.  I was not planning on giving transformed parameters access to the "generated data/latent data/cut parameters/..."

Interesting idea about re-using the transformed parameters block, as that'd neatly solve the problem.  We'd thought about merging with generated quantities, but that doesn't make so much sense.

Reply:
Can you not have continuous parameters generated in the latent data block? I like `latent data`

Reply:
Yes, I'll add a statement to make it clear that any types, including constrained types, may be declared.

Reply:
Topic 1:  `cut parameters` name:  I added a discussion to design doc of the choices so far.  So far, `cut parameters` is my personal favorite, but I'm leaving `latent data` in the design doc for now.

Topic 2:  Ordering of transformed parameters and cut parameters---I added this to the unresolved questions section.  Putting the cut parameters first is more flexible, but may require duplicated work if they depend on transformed parameters.  (This is why we want a blockless Stan!)

Topic 3:  I made sure to cliarfy that there can be continuous parameters in the cut parameters block.

With that, I'm going to mark this resolved, though there's still the outstanding issue of ordering.